`Typography` doc enhancements - `Website` (#2364)
hashicorp/design-system: website/docs/foundations/typography/partials/guidelines/guidelines.md

## How typography is referenced

> The construction of typographic styles varies heavily depending on context and tooling. This overview will clarify how HDS typography standards are implemented in different tools.

Reviewer: Maybe you can reduce the sentence length and complexity here?
```suggestion
Typography styles change based on the tool. This overview explains how HDS typography standards work in each tool.
```

In Figma, reusable text styles are called “Text Styles” and are categorized and stored in the “text panel” to be used in text-based elements. Text styles directly reference font family, weight, line height, size, and many other text-based properties within a single style and are manually maintained. Some Figma styles include:

- `Display/500/Bold`
- `Body/300/Regular`
- `Code/300/Regular`

In code, these styles are constructed differently, using “Tokens.” Tokens are generated from an agnostic JSON file and differ from Figma “Text Styles” because a single token does not reference multiple font properties. Instead, each Token defines the different properties of a single text style. This means that constructing the equivalence of Figma’s Text Style `Display/500/Bold` would utilize the following tokens:

- `font-family: var(--token-typography-display-500-font-family);`
- `font-size:var(--token-typography-display-500-font-size);`
- `line-height:var(--token-typography-display-500-line-height);`
- `font-weight:var(--token-typography-font-weight-bold);`

> Rather than calling upon each of these tokens in code every time text is styled, the [Text Component](/components/text) was created to replicate how Figma styles group text-based properties.

Author: I'll reference this, thank you 🙏

## Font families

> Helios uses the default system fonts available on each OS. These system fonts are stable and support internationalization, code, tabular data, data visualization, etc. Additionally, they improve performance by reducing the number of resources to be loaded, which is important for large applications like those at HashiCorp.

Reviewer: To Jorys suggestion, but also maybe shortened and using a more active voice?

```suggestion
In code, Helios uses default system fonts because they are stable, support internationalization, and improve performance by reducing resource load—key for large applications at HashiCorp.
```

While text styling varies slightly across operating systems, users will have a consistent and expected experience within each OS.

### Sans-serif

> ![sans-serif on mac](/assets/foundations/typography/sans-serif-mac.png =660x*)

Reviewer 2: ```suggestion
![The sans-serif font on Mac is SF Pro Display for 20px and above and SF Pro Text for 19px and below.](/assets/foundations/typography/sans-serif-mac.png =660x*)
```

Reviewer 2: Suggestion: could be out of scope for this PR, but we could rethink these screenshots to move some of the text content out of the image and into the HTML of the page. This would  benefit users because long alt text is not recommended from an a11y perspective and the text would be searchable.

Reviewer 2: an alternate design could be like this (ignore my hacky Figma usage lol)
![Screenshot 2024-08-27 at 5 27 23 PM](https://github.com/user-attachments/assets/f5752706-fe22-4bf7-a657-fdcc8efc7ed2)

Reviewer 3: I agree that would would be better to include less of the text within the images such as the labels for the particular operating system.

Author: This is a good idea!

> ![sans-serif on windows](/assets/foundations/typography/sans-serif-windows.png =660x*)

Reviewer 2: ```suggestion
![The sans-serif font on Windows is Segoe Ul Display for 20px and above and Segoe Ul Text for 19px and below.](/assets/foundations/typography/sans-serif-windows.png =660x*)
```

> ![sans-serif on linux](/assets/foundations/typography/sans-serif-linux.png =660x*)

Reviewer 2: ```suggestion
![The sans-serif font for Linux is the system sans-serif font, which depends on the Linux distribution.](/assets/foundations/typography/sans-serif-linux.png =660x*)
```

### Monospace

> ![monospace on mac](/assets/foundations/typography/monospace-mac.png =660x*)

Reviewer 2: Same feedback for all the monospace images as the san serif ones.

![monospace on windows](/assets/foundations/typography/monospace-windows.png =660x*)

![monospace on linux](/assets/foundations/typography/monospace-linux.png =660x*)

## Font weights

In addition to font size, font weight helps create different levels of emphasis throughout the UI. Font weights include:

- 400 (regular)
- 500 (medium)
- 600 (semi bold)
- 700 (bold)

Within these ranges, bold is used only for display and some code text styles, while regular is associated with body text styles which may include code.

## Text styles

### Display

> Display styles should be used for headings or hierarchical emphasis (like remaining balances). Unless used for hierarchical emphasis, if the display text looks like a page heading, it should use an [HTML heading tag](https://www.w3.org/WAI/WCAG22/Understanding/section-headings.html).

Reviewer 4: This might also be a place where you can jump link users to the "Semantic heading or not" section that you have at the end of the page

> Display styles should be used for headings or hierarchical emphasis (like remaining balances). Unless used for hierarchical emphasis, if the display text looks like a page heading, it should use an [HTML heading tag](https://www.w3.org/WAI/WCAG22/Understanding/section-headings.html).

Reviewer 2: I think im confused by the phrase "hierarchical emphasis"... could it be "visual emphasis"? although I don't really like that either.. its really a place to attract visual attention when there is no semantic reason to.

> Display styles should be used for headings or hierarchical emphasis (like remaining balances). Unless used for hierarchical emphasis, if the display text looks like a page heading, it should use an [HTML heading tag](https://www.w3.org/WAI/WCAG22/Understanding/section-headings.html).

Author: Maybe it should just be visual emphasis? Because thats the best way to describe this.

Display text styles are available in the following sizes and weights:

> ![Display 500 at 30px font size, with bold font weight. Display 400 at 24px font size, with bold, semibold and medium font weights. Display 300 at 18px font size, with bold, semibold and medium font weight. Display 200 at 16px font size, with semibold font weight. Display 100 at 13px font size, with medium font weight.](/assets/foundations/typography/typography-display-options.png)

Reviewer 2: This is another situation where I would encourage moving the text out of an image and into code. Alt text doesn't support rich text like lists, so this is very hard to parse.

The text in the image could literally just be items in an HTML list that we style to not have bullets and each list item to have the correct display styles. 

```html
<ul>
  <li>Display 500 (30px) - Bold</li>
  <li>Display 400 (24px) - Bold</li>
  <li>Display 400 (24px) - Semibold</li>
  <li>Display 400 (24px) - Medium</li>
  // etc.
</ul>
```

Reviewer 2: This goes for the body and code sections as well.

Body text styles are available in the following weights and sizes:

![Body 300 at 16px font size, with semibold, medium and regular font weights. Body 200 at 14px font size, with semibold, medium and regular font weights. Body 100 at 13px font size, with semibold, medium and regular font weights.](/assets/foundations/typography/typography-body-options.png)

### Code

Code styles are used exclusively for text representing code in the UI (like the [Code Block component](/components/code-block)). 

Code text styles come in the following weights and sizes:

![Code 300 at 16px font size, with bold and regular font weights. Code 200 at 14px font size, with bold and regular font weights. Bode 100 at 13px font size, with bold and regular font weights.](/assets/foundations/typography/typography-code-options.png)

## How typography is used

### Body content

> Body 200 is the default font size for body content. Scaling up or down is used as a means to create visual hierarchy in page UI, or component usage. As an example, Helio’s `Button`, `medium` (default) variant is sized at Body 200 while the `large` variant is Body 300.

Reviewer 5: [suggestion]
```suggestion
Body 200 is the default font size for body content. Scaling up or down is used as a means to create visual hierarchy in page UI, or component usage. As an example, the Button `medium` (default) variant is sized at Body 200 while the `large` variant is Body 300.
```

![Two buttons, one medium and one large](/assets/foundations/typography/typography-button-example.png)

> Similarly, form fields’ labels and input texts are Body 200, while any other secondary information, like helper text or error messages are sized at Body 100. This ensures that from a hierarchy perspective, the primary information is more prominent with the larger size of the two.

Reviewer 3: ```suggestion
Similarly, form field labels and input texts use Body 200, while any other secondary information, like helper text or error messages are sized at Body 100. This ensures that from a hierarchy perspective, the primary information is more prominent.
```

> Similarly, form fields’ labels and input texts are Body 200, while any other secondary information, like helper text or error messages are sized at Body 100. This ensures that from a hierarchy perspective, the primary information is more prominent with the larger size of the two.

Reviewer 4: ```suggestion
Similarly, form fields’ labels and input texts are Body 200, while any other secondary information, like helper text or error messages are sized at Body 100. This ensures that primary information is more prominent with the larger size for visual hierarchy.
```

> ![A form field set with label, helper text an input field errored out along with an error message.](/assets/foundations/typography/typography-form-example.png)

Reviewer 3: ```suggestion
![A form field with label, helper text, an input field in an error state, and an error message.](/assets/foundations/typography/typography-form-example.png)
```

> For more details about form patterns, read the [form patterns](/patterns/form-patterns) documentation for additional details.

Reviewer 4: ```suggestion
For more details about form patterns, read the [form patterns](/patterns/form-patterns) documentation.
```

### Semantic heading or not?

> Even though Helio’s Text Component does not equate a specific style (Body, Display, or Code) with a specific HTML semantic structure (like `<p/>`, `<h1/>` or `<span>`), this is up to the consumer to define the logical order in which the text is being displayed as. To ensure accessible page content, the scaling of the Display styles often are associated with the scaling of the HTML headings such as:

Reviewer 3: ```suggestion
As Helio’s Text component does not equate to a specific style (Body, Display, or Code) or a semantic HTML element (like `<h1/>`, `<h2/>`, etc.), it is up to the consumer to define the tag with the appropriate sequential order for the heading text. To ensure accessible page content, the scaling of the Display styles are often associated with the scaling of the HTML headings such as:
```

Reviewer 3: This section is related to appropriate use of header styles & HTML markup while the text was more general so seemed confusing to me. (So I made the above suggested edits)

Reviewer 5: [suggestion] I would also remove the reference to Helios (redundant)
```suggestion
Even though the [Text](/components/text) component does not equate a specific style (Body, Display, or Code) with a specific HTML semantic structure (like `<p>`, `<h1>` or `<span>`), this is up to the consumer to define the logical order in which the text is being displayed as. To ensure accessible page content, the scaling of the Display styles often are associated with the scaling of the HTML headings such as:
```

Reviewer 2: I would even be a little more forceful with this: 
```suggestion
The Helio’s Text Component does not equate a specific style (Body, Display, or Code) with a specific HTML semantic structure (like `<p/>`, `<h1/>` or `<span>`), it is up to the consumer to define the logical semantic element for the each piece of text. To ensure accessible page content, the scaling of the Display styles often are associated with the scaling of the HTML headings such as:
```

> - Display 500 -> H1

Reviewer 5: [thought] I am worried that this "strict" association may lead to misunderstandings more than being useful (for the reasons explained right above)

Reviewer 6: Yeah, maybe this would benefit from a bit of a disclaimer, like "In general we recommend mapping Display font sizes sequentially to markup, but this is not always possible."

Author: That's fair. I'll add this note.

- Display 400 -> H2
- Display 300 -> H3
- Display 200 -> H4
- Display 100 -> H5

> However, there are instances where a Display text style will be used as a means to communicate something of value that is not a page heading. This is used to emphasize an important piece of information, such as remaining balances, so that users are able to visually understand the level of importance of this information on the page.

Reviewer 4: ```suggestion
However, Display test styles do not always need to be headings. They may be used to emphasize important information, such as remaining balances, so that users are able to visually understand the level of importance of information on the page.
```

> ![A piece of UI showing primarily an account balance, highlighting the differences between the two display styles. One a heading and the other one not.](/assets/foundations/typography/typography-remaining-balance-example.png)

Reviewer 2: Maybe @MelSumner  has a diff opinion than me, but I would say in this example the balance is a part of the H3 - just broken onto a separate line.

Author: In this instance, we may want to lean into it not being an `H3` for the example's purpose, unless this outright fails on accessibility (which I don't believe so)?

Reviewer 2: Yeah its fine.

- `font-family: var(--token-typography-display-500-font-family);`
- `font-size:var(--token-typography-display-500-font-size);`
- `line-height:var(--token-typography-display-500-line-height);`
- `font-weight:var(--token-typography-font-weight-bold);`

Rather than calling upon each of these tokens in code every time text is styled, the [Text Component](/components/text) was created to replicate how Figma styles group text-based properties. Additional [CSS helper classes](https://helios.hashicorp.design/foundations/typography?tab=code#css-helper-classes) have been created to ease the style creation process, if necessary.

## Font families

In code, Helios uses default system fonts because they are stable, support internationalization, and improve performance by reducing resource load—key for large applications at HashiCorp.

In Figma, our design tooling and libraries use SF Pro font family exclusively and we recommend using this font stack for consistency.

While text styling varies slightly across operating systems, users will have a consistent and expected experience within each OS.

### Sans-serif

#### Mac + iOS sans-serif font

Large text (20px and above): SF Pro Display

> ![Sans-serif display on Mac](/assets/foundations/typography/typography-sans-serif-display-mac.png)

Reviewer 2: I should have said this in my previous round of comments, but now that the informational content is moved out of the images I think we can make all the font sample images decorative.

```suggestion
![](/assets/foundations/typography/typography-sans-serif-display-mac.png)
```

#### Windows

Consolas

![Monospace on Mac](/assets/foundations/typography/typography-monospace-menlo-mac.png)

#### Linux

Linux monospace system font dependant on the Linux distribution

![Monospace on Linux](/assets/foundations/typography/typography-monospace-linux.png)

## Font weights

In addition to font size, font weight helps create different levels of emphasis throughout the UI. Font weights include:

- 400 (regular)
- 500 (medium)
- 600 (semi bold)
- 700 (bold)

Within these ranges, bold is used only for display and some code text styles, while regular is associated with body text styles which may include code.

## Text styles

### Display

Display styles should be used for headings or visual emphasis (like remaining balances). Unless used for visual emphasis, if the display text looks like a page heading, it should use an [HTML heading tag](https://www.w3.org/WAI/WCAG22/Understanding/section-headings.html).

Display text styles are available in the following sizes and weights:

<Hds::Text::Display @size="500" @weight="bold">Display 500 (30px) - Bold</Hds::Text::Display>

> <br/>

Reviewer 2: Because this is presented as a list, I really think it should have the list semantics. The benefit is that users can know how many items there are.

Ideally, you use the semantic list HTML elements and add styles to remove the bullet points:
```handlebars
<ul class='doc-typography-demo-display-list'>
  <Hds::Text::Display @tag="li" @size="500" @weight="bold">Display 500 (30px) - Bold</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display @tag="li" @size="400" @weight="bold">Display 400 (24px) - Bold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display @tag="li" @size="400" @weight="semibold">Display 400 (24px) - Semibold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display @tag="li" @size="400" @weight="medium">Display 400 (24px) - Medium</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display @tag="li" @size="300" @weight="bold">Display 300 (18px) - Bold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display @tag="li" @size="300" @weight="semibold">Display 300 (18px) - Semibold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display @tag="li" @size="300" @weight="medium">Display 300 (18px) - Medium</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display @tag="li" @size="200" @weight="semibold">Display 200 (16px) - Semibold</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display @tag="li" @size="100" @weight="medium">Display 100 (13px) - Medium</Hds::Text::Display>
</ul>
```

or you can manually set the roles
```handlebars
<div role='list'>
  <Hds::Text::Display role='listitem' @size="500" @weight="bold">Display 500 (30px) - Bold</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display role='listitem' @size="400" @weight="bold">Display 400 (24px) - Bold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display role='listitem' @size="400" @weight="semibold">Display 400 (24px) - Semibold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display  role='listitem' @size="400" @weight="medium">Display 400 (24px) - Medium</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display role='listitem' @size="300" @weight="bold">Display 300 (18px) - Bold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display role='listitem' @size="300" @weight="semibold">Display 300 (18px) - Semibold</Hds::Text::Display>
  <br/>
  <Hds::Text::Display role='listitem' @size="300" @weight="medium">Display 300 (18px) - Medium</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display role='listitem' @size="200" @weight="semibold">Display 200 (16px) - Semibold</Hds::Text::Display>
  <br/>
  <br/>
  <Hds::Text::Display role='listitem' @size="100" @weight="medium">Display 100 (13px) - Medium</Hds::Text::Display>
</div>
```

Reviewer 2: Same for body and code.

Reviewer 5: In this case, following @shleewhite suggeston, consider using plain HTML (`<ul>` + `<li>`) with CSS typographic class names (it would make the code much easier to interpret). See how it's done here: https://helios.hashicorp.design/foundations/typography?tab=code#style-and-weight (eg. `hds-typography-display-500 hds-font-weight-bold` for "Display/500/bold"

![Two buttons, one medium and one large](/assets/foundations/typography/typography-button-example.png)

Similarly, form fields' labels and input texts are Body 200, while any other secondary information, like helper text or error messages are sized at Body 100. This ensures that primary information is more prominent with the larger size for visual hierarchy.

![A form field with label, helper text, an input field in an error state, and an error message.](/assets/foundations/typography/typography-form-example.png)

For more details about form patterns, read the [form patterns](/patterns/form-patterns) documentation.

### Semantic heading or not?

The [Text](/components/text) component does not equate a specific style (Body, Display, or Code) with a specific HTML semantic structure (like `<p/>`, `<h1/>` or `<span>`), it is up to the consumer to define the logical semantic element for the each piece of text. To ensure accessible page content, the scaling of the Display styles often are associated with the scaling of the HTML headings such as:

> - Display 500 -> H1

Reviewer 2: Writing an arrow this way doesn't get announced as you might expect with a screen reader. VoiceOver ignores the "-" and just says "Display 500 greater than H1". There is an ascii character for right arrow that announces correctly that you can use instead.

```suggestion
- Display 500 → H1
```

Author: Oh wow, great to know, thanks for the education :D
